bluepill/src/BPHTMLReportWriter.m

```diff
@@ -0,0 +1,160 @@
+//  Copyright 2016 LinkedIn Corporation
+//  Licensed under the BSD 2-Clause License (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at https://opensource.org/licenses/BSD-2-Clause
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OF ANY KIND, either express or implied.  See the License for the specific language governing permissions and limitations under the License.
+
+#import <Foundation/Foundation.h>
+#import "BPTestReportHTML.h"
+#import "bp/src/BPUtils.h"
+
+#import "BPHTMLReportWriter.h"
+
+static const NSString * const kHTMLPath = @"0-test-report.html";
+static const NSString * const kJSPath = @"test-report.js";
+static const NSString * const kFailureLogsPath = @"failure-logs";
+
+@implementation BPHTMLReportWriter
+
+- (NSDictionary<NSString*, NSString*> *)metaForFileWrites:(NSXMLDocument *)jUnitReport
+                                                 inFolder:(nonnull NSString *)folderPath {
+    NSMutableDictionary *metaData = [[NSMutableDictionary alloc] init];
+    NSString *htmlPath = [folderPath stringByAppendingPathComponent:[kHTMLPath copy]];
+    NSString *jsPath = [folderPath stringByAppendingPathComponent:[kJSPath copy]];
+    NSString *failureLogsPath = [folderPath stringByAppendingPathComponent:[kFailureLogsPath copy]];
+    // HTML data
+    metaData[htmlPath] = kHTMLContent;
+    [BPUtils printInfo:INFO withString:@"HTML report: %@", htmlPath];
+
+    // write the JS goes with the html
+    if (jUnitReport.childCount == 0) {
+        [BPUtils printInfo:WARNING withString:@"Bluepill JUnit report has no data."];
+        return nil;
+    }
+    NSMutableArray<NSString *> *jsLines = [[NSMutableArray alloc] initWithArray:@[
+                                                                                  @"var json = {",
+                                                                                  @"\t\"product\": \"All Tests\",\n",
+                                                                                  @"\t\"testSuites\": [",
+                                                                                  @"\t\t{",
+                                                                                  @"\t\t\t\"testCases\": ["
+                                                                                  ]];
+    // testsuites
+    // <testsuites name="AllTestUnits" tests="17" failures="0" errors="1" time="53.990912">
+    NSXMLNode *rootNode = jUnitReport.children[0];
```

Comment:
we should invest on understanding those test case start/finish call backs, parsing XMLs manually doesn't really look clean .. esp. as bluepill is becoming a LTS tool :D

Reply:
We are not parsing the XML again but we are consuming the JUnit report in memory before it's written to disk. The JUnit report already has that covered.